### DISCUSSION TITLE
CTE optimization via WITH TAIL RECURSIVE

### DISCUSSION THREAD

**Discussion by pkoppstein** [duckdb/duckdb#14447] [Ideas]
Currently DuckDB's support for recursion seems to be limited to recursive CTEs,
but they are not optimized for one of the typical use cases - in which essentially only
the most recently added line ever needs to be retained.

The program shown below is representative of this class, though it is of course
useless except for making the point that the failure to optimize seems
to show up both in terms of speed and memory usage.  Here are
some performance metrics obtained using `/usr/bin/time -lp` with
DuckDB v1.1.2-dev218:
```
                                n=1             1 million
user                           0.03             98.03s
sys                            0.00             28.65s
maximum resident set size  11726848           17186816  
cycles elapsed             13616695       406474522801  
peak memory footprint       2842624            8282112  
```
The mrse and peak memory footprint increased by about 5MB, in line with what one
would expect of an unoptimized implementation.

Ideally, DuckDB would recognize when only the most-recently added row is relevant (in the present cast, the big hint is the use of last()), but in general
that might be difficult to determine, and hence the title of this ER,
suggesting a new keyword to allow the user to declare that DuckDB can
optimize the CTE on the assurance that only the last line, once constructed, need
be retained during the iteration.

One would like to think that the run-time would also be drastically reduced
if this optimization were implemented.
```
create or replace function i(n) as table (
  with recursive cte as (
  select 1 as i
    union all
    select i + 1 as i
    from cte
    where i < n )
  select last(i) from cte
);


# from i(1);
from i(1000000);
```

**Comment by soerenwolfers** [comment]
wouldn't a regular recursive function be better suited for that use case?

**Comment by pkoppstein** [comment]
@soerenwolfers asked:
> wouldn't a regular recursive function be better suited for that use case?

DuckDB's "MACROs" seem quite far removed from conventional programming-language functions, let alone recursive functions, let alone recursive functions with tail-recursive optimization, whereas DuckDB's recursive CTEs are almost there, and I am hoping that adding a compiler hint (e.g. "TAIL") will nudge DuckDB over the finishing line.

**Comment by billwallis** [comment]
This is something that I'd love to have, too 🤓 

I've had cases where I want to keep applying a rule until a condition is met, without needing to keep the iteration history; for these, I'll usually have some iteration counter (say, `i`) and I'll just keep the rows from the recursive CTE whose iteration counter is the last iteration counter:

```sql
with recursive loop(x, i) as (
        select random(), 0
    union all
        select
            x + 0.01 as x_,
            i + 1,
        from loop
        where x_ < 1
)

from loop
qualify i = max(i) over ()
```

In cases like this, keeping all the intermediate rows is redundant and, depending on the scenario, can gobble up a lot of memory -- so it'd be awesome to be able to tell a recursive CTE to only keep the latest intermediate result rather than append it to the previous intermediate results

If what I'm describing matches what @pkoppstein has asked for (I _think_ it does), then I was imagining a keyword like `shallow` which would fit where [`materialized`](https://duckdb.org/docs/sql/query_syntax/with.html#cte-materialization) sits in the syntax tree. I'd also be happy with the proposed `tail` instead of `shallow`

With this, I'd expect to be able to write something like:

```sql
with recursive loop(n, i) as shallow (
        values (2, 0), (3, 0)
    union all
        select n * 2, i + 1,
        from loop
        where i < 5
)

from loop
```

...whose result set would be:

```
┌─────────┬─────────┐
│    n    │    i    │
│  int32  │  int32  │
├─────────┼─────────┤
│     64  │      5  │
│     96  │      5  │
└─────────┴─────────┘
```

This is a silly example and is just for illustration -- and I used two rows to be clear that I'd want the last intermediate result set, not just the last row

---

@pkoppstein if this is different to what you're asking for here, I'll move this comment into a separate discussion
